packages/asset-swapper/contracts/src/MakerPSMSampler.sol

```diff
@@ -0,0 +1,251 @@
+                return 0;
+            }
+
+            uint256 feeInWad = takerTokenAmountInWad.safeMul(psm.tin()).safeDiv(WAD);
+            uint256 makerTokenAmountInWad = takerTokenAmountInWad.safeSub(feeInWad);
+
+            return makerTokenAmountInWad;
+        } else if (makerToken == psmInfo.gemTokenAddress) {
+            // Simulate buyGem
+            // Buying USDC from the PSM, decreasing the total debt
+            // Selling DAI for USDC, already in 18 decimals [wad]
+            uint256 takerTokenAmountInWad = takerTokenAmount;
+            uint256 newTotalDebtInRad = totalDebtInWad.safeSub(takerTokenAmountInWad).safeMul(RAY);
+
+            // PSM is empty, not enough USDC to buy from it
+            if (newTotalDebtInRad <= debtFloorInRad) {
+                return 0;
+            }
+
+            uint256 feeDivisorInWad = WAD.safeAdd(psm.tout()); // eg. 1.001 * 10 ** 18 with 0.1% tout;
+            uint256 makerTokenAmountInGemTokenBaseUnits =  takerTokenAmountInWad.safeMul(gemTokenBaseUnit).safeDiv(feeDivisorInWad);
+
+            return makerTokenAmountInGemTokenBaseUnits;
+        }
+
+        return 0;
+    }
+
+    function _samplePSMBuy(MakerPsmInfo memory psmInfo, address makerToken, address takerToken, uint256 makerTokenAmount, IPSM psm, IVAT vat)
+        private
+        view
+        returns (uint256)
+    {
+        (uint256 totalDebtInWad,,, uint256 debtCeilingInRad, uint256 debtFloorInRad) = vat.ilks(psmInfo.ilkIdentifier);
+
+        if (takerToken == psmInfo.gemTokenAddress) {
+            // Simulate sellGem
+            // Selling USDC to the PSM, increasing the total debt
+            uint256 makerTokenAmountInWad = makerTokenAmount;
+            uint256 feeDivisorInWad = WAD.safeSub(psm.tin()); // eg. 0.999 * 10 ** 18 with 0.1% tin;
+            uint256 takerTokenAmountInWad = makerTokenAmountInWad.safeMul(WAD).safeDiv(feeDivisorInWad);
+            uint256 newTotalDebtInRad = totalDebtInWad.safeAdd(takerTokenAmountInWad).safeMul(RAY);
+
+            // PSM is too full to fit
+            if (newTotalDebtInRad >= debtCeilingInRad) {
+                return 0;
+            }
+
+            uint256 takerTokenAmountInGemInGemBaseUnits = (takerTokenAmountInWad.safeDiv(1e12)).safeAdd(1); // Add 1 to deal with cut off decimals converting to lower decimals
+
+            return takerTokenAmountInGemInGemBaseUnits;
+        } else if (makerToken == psmInfo.gemTokenAddress) {
+            // Simulate buyGem
+            // Buying USDC from the PSM, decreasing the total debt
+            uint256 makerTokenAmountInWad = makerTokenAmount.safeMul(1e12);
+            uint256 feeMultiplierInWad = WAD.safeAdd(psm.tout()); // eg. 1.001 * 10 ** 18 with 0.1% tout;
+            uint256 takerTokenAmountInWad =  makerTokenAmountInWad.safeMul(feeMultiplierInWad).safeDiv(WAD);
+            uint256 newTotalDebtInRad = totalDebtInWad.safeSub(takerTokenAmountInWad).safeMul(RAY);
```

Comment:
same Q here

Reply:
Yep, should add a guard for that here too 🙏